server/sonar-server/src/main/java/org/sonar/server/issue/ws/ChangelogAction.java

```diff
@@ -0,0 +1,157 @@
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Stream.of(request)
+        .map(searchChangelogs(dbSession))
+        .map(buildResponse())
+        // TODO use Collectors.toOneElement to improve readability
+        .forEach(wsResponse -> writeProtobuf(wsResponse, request, response));
+    }
+  }
+
+  private Function<Request, ChangeLogResults> searchChangelogs(DbSession dbSession) {
+    return request -> new ChangeLogResults(dbSession, request.mandatoryParam(PARAM_ISSUE));
+  }
+
+  private static Function<ChangeLogResults, ChangelogWsResponse> buildResponse2() {
+    return result -> ChangelogWsResponse.newBuilder().addAllChangelog(
+      result.changes.stream()
+        .map(toWsChangelog(result))
+        .collect(Collectors.toList()))
+      .build();
+  }
+
+  private static Function<ChangeLogResults, ChangelogWsResponse> buildResponse() {
+    return result -> Stream.of(ChangelogWsResponse.newBuilder())
+      .peek(addChanges(result))
+      .map(ChangelogWsResponse.Builder::build)
+      // TODO replace by Collectors.toOneElement
+      .findAny().orElseThrow(IllegalStateException::new);
+  }
+
+  private static Consumer<ChangelogWsResponse.Builder> addChanges(ChangeLogResults results) {
+    return response -> results.changes.stream()
+      .map(toWsChangelog(results))
+      .forEach(response::addChangelog);
+  }
+
+  private static Function<FieldDiffs, Changelog> toWsChangelog(ChangeLogResults results) {
+    return change -> {
+      String userLogin = change.userLogin();
+      Changelog.Builder changelogBuilder = Changelog.newBuilder();
+      changelogBuilder.setCreationDate(formatDateTime(change.creationDate()));
+      UserDto user = userLogin == null ? null : results.users.get(userLogin);
+      if (user != null) {
+        changelogBuilder.setUser(user.getLogin());
+        changelogBuilder.setUserName(user.getName());
+        changelogBuilder.setEmail(user.getEmail());
+      }
+      change.diffs().entrySet().stream()
+        .map(toWsDiff())
+        .forEach(changelogBuilder::addDiffs);
+      return changelogBuilder.build();
+    };
+  }
+
+  private static Function<Map.Entry<String, FieldDiffs.Diff>, Changelog.Diff> toWsDiff() {
+    return diff -> {
+      FieldDiffs.Diff value = diff.getValue();
+      Changelog.Diff.Builder diffBuilder = Changelog.Diff.newBuilder();
+      String key = diff.getKey();
+      diffBuilder.setKey(key.equals(TECHNICAL_DEBT) ? EFFORT_CHANGELOG_KEY : key);
+      setNullable(emptyToNull(value.newValue().toString()), diffBuilder::setNewValue);
+      setNullable(emptyToNull(value.oldValue().toString()), diffBuilder::setOldValue);
+      return diffBuilder.build();
+    };
+  }
+
+  private class ChangeLogResults {
+    private final List<FieldDiffs> changes;
+    private final Map<String, UserDto> users;
+
+    // TODO I'd separate the search and the content of the results
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)